### DISCUSSION TITLE
Daisy chained PCIe switches?

### DISCUSSION THREAD

**Discussion by bryanvaz** [geerlingguy/raspberry-pi-pcie-devices#378] [Ideas]
Thought of a hilarious idea while shopping for Pi accessories listening to the new USB video of yours: have you ever tried daisy chain your PCIe switches?

I know TI PCIe fabric switches can be daisy chained, but I don't know if any of the other vendors, like Microsemi, baked the functionality into their stack. 

Also for laughs someone tried to patent daisy chaining PCIe devices back in 2012; <<insert comment about pointless patents here>> (https://patents.google.com/patent/US20130179621A1/en).

**Comment by geerlingguy** [comment]
I actually did one time, never posted anything on it though. Technically it works, but some things like NVMe functionality can be funky once you have more than one switch in the chain—at least on Pi OS.

**Comment by 6by9** [comment]
Yes it works as long as you have patch https://github.com/raspberrypi/linux/commit/a514fef1c808b2fac84a4d03b16d93d9a6d78ffc (also in mainline Linux kernels).

I had an ASM1184 feeding a Pericom PI7C9X whilst testing for #57, and the Hauppauge WinTV QuadHD is 2 dual tuners behind a PCIe switch which I've had working on the ASM1184.

**Comment by bryanvaz** [comment]
Ok now you two have definitely piqued my interest.

I think I'll try pairing an Alftel 12x card with an ASM1184e board (like the ASHATA PCE4PCE-A01), and try two ASM1184e in a chain as a backup. If only the Coral worked properly with the Pi out of the box, that would be the perfect use case! 

As far as NVMe, I remember in large NVMe subsystems you need retimers in the chain otherwise drives start to drop out because the protocol is so latency sensitive. I'm guessing both of you haven't had a chance to try a retimer card with the Pi?
